File: Master_DSA_Practices/SESSION_05/sorting.cpp
#include <iostream> 
#include <stdexcept>
#include <cstdlib> 
#include <ctime> 

using ssize_t = long long; 

class array
{
    private: 
        int* a; 
        ssize_t N; 

        static void insert_at_sorting_position(int* a, ssize_t N)
        {
            int key;
            ssize_t i; 

            key = a[N-1]; 
            for(i = N-2; i > -1 && a[i] > key; --i)
                a[i+1] = a[i]; 
            a[i+1] = key; 
        }

    public: 
        array(ssize_t _N) 
        {
            if(_N <= 0)
                throw std::domain_error("Bad size for array"); 

            a = new int[_N]; 
            N = _N; 
        }

        void fill_array()
        {
            srand(time(0)); 
            for(ssize_t i = 0; i < N; ++i)
                a[i] = rand(); 
        }

        void sort() 
        {
            for(ssize_t n = 2; n <= N; ++n)
                insert_at_sorting_position(a, n); 
        }

        friend std::ostream& operator<<(std::ostream& os, const array& A)
        {
            for(ssize_t i = 0; i < A.N; ++i)
                os << "a[" << i << "]:" << A.a[i] << std::endl; 
            return os;  
        }

        ~array()
        {
            delete[] a; 
            a = nullptr; 
        }

 
        array(const array& other) = delete; 
        array(const array&& other) = delete; 
        array& operator=(const array& other) = delete; 
        array& operator=(const array&& other) = delete; 
    
}; 

int main(int argc, char* argv[])
{
    ssize_t N; 

    if(argc != 2)
    {
        std::cerr << "Usage Error:" << argv[0] << " Number of elements" << std::endl; 
        exit(EXIT_FAILURE); 
    }

    N = atoll(argv[1]); 
    if(N <= 0)
    {
        std::cerr << "Bad size for array" << std::endl; 
        exit(EXIT_FAILURE); 
    }

    array* pA = new array(N); 
    pA->fill_array(); 
    
    std::cout << "Array Before Sort" << std::endl << *pA; 
    pA->sort(); 
    std::cout << "Array After Sort" << std::endl << *pA; 

    delete pA; 
    pA = nullptr; 

    return (EXIT_SUCCESS); 
}

File: Master_DSA_Practices/SESSION_05/sorting.py
import random, sys

class Array_iterator: 
    def __init__(self, G): 
        self.G = G 
    def __next__(self): 
        return self.G.__next__()
        
class Array: 
    def __init__(self, N:int):
        if type(N) != int: 
            raise TypeError("Bad type for int")
        if N <= 0: 
            raise ValueError("Bad value for array size") 
        self.a = [random.randint(0, 100000) for n in range(N)]

    def sort(self) -> None: 
        for j in range(1, len(self.a)): 
            key = self.a[j]
            i = j - 1
            while i > -1 and self.a[i] > key: 
                self.a[i+1] = self.a[i]
                i = i - 1 
            self.a[i+1] = key 

    def __iter__(self): 
        def get_generator(L:list): 
            for x in L:
                yield x 
            return Array_iterator(get_generator(self.a))

    def __getitem__(self, index_or_slice): 
        if type(index_or_slice) == int: 
            return self.a[index_or_slice]
        tmpA = Array(1)
        tmpA.a = self.a[index_or_slice] 
        return tmpA 

    def __len__(self): 
        return len(self.a)

def main(argc, argv): 
    if argc != 2: 
        print("Usage Error: %s number of elements" % argv[0])
        sys.exit(-1)
    
    A = Array(int(argv[1]))

    print("Before sort:")
    for i in range(len(A)): 
        print("A[{}]:{}".format(i, A[i]))

    A.sort()

    print("After sort:")
    for i in range(len(A)): 
        print("A[{}]:{}".format(i, A[i]))

    sys.exit(0) 

main(len(sys.argv), sys.argv)

